src/source.rs

```diff
@@ -379,3 +379,93 @@ struct Artifact {
     location: String,
     signature: String,
 }
+
+pub trait IgnitionLocation: Display {
+    /// Verify that the Ignition location is valid
+    fn check(&self) -> Result<()>;
+    /// Start fetching the Ignition config
+    fn read(&self) -> Result<Box<dyn Read>>;
+}
+
+/// Local Ignition config
+#[derive(Debug)]
+pub struct IgnitionFileLocation(PathBuf);
+
+/// Remote Ignition config via URL
+#[derive(Debug)]
+pub struct IgnitionUrlLocation(Url);
+
+impl IgnitionFileLocation {
+    fn read(&self) -> Result<std::fs::File> {
```

Comment:
Confusing to have a trait and struct method with the same name.  Why not just have this code in the trait `read()`?

Reply:
Yeah thanks, it's confusing. I renamed it to `open_file()`, which then made me displeased with the name `read()`, which I just changed to `open()`.

It's not in the trait implementation because it's used in both `check()` and `open()`.

Reply:
Right, but the `check()` trait method can call the trait `read()`.

Reply:
Right, but then we're allocating memory for the `Box` when all we need is `open(2)`.